### ISSUE TITLE
Generated mapping causes Stackoverflow

### ISSUE THREAD

**Issue by AldrichMascarenhas** [mapstruct/mapstruct#1730]
```java
 import com.example.core.data.database.model.ContentType
 import org.mapstruct.*
 import com.example.core.data.network.Content as nwContent
 import com.example.core.data.database.model.Content as dbContent
 import org.mapstruct.MappingTarget
 import org.mapstruct.AfterMapping
 
 @Mapper
 interface ContentMapper {
 
     @Mapping(target = "contentType", ignore = true)
     fun convertToMovieDbContent(content: nwContent, @Context contentType: ContentType): dbContent
     
     @AfterMapping
     fun setContentType(content: nwContent, @MappingTarget dbContent: dbContent, @Context contentType: ContentType) {
         dbContent.contentType = contentType
     }
     
 }
 
```
Generated class: 
 
```java
 package com.example.core.data.mapper;
 
 import com.example.core.data.database.model.ContentType;
 import com.example.core.data.network.Content;
 import java.util.ArrayList;
 import java.util.List;
 
 /*
 @Generated(
     value = "org.mapstruct.ap.MappingProcessor",
     date = "2019-02-20T17:43:28+0000",
     comments = "version: 1.3.0.Final, compiler: javac, environment: Java 1.8.0_152-release (JetBrains s.r.o)"
 )
 */
 public class ContentMapperImpl implements ContentMapper {
 
     @Override
     public com.example.core.data.database.model.Content convertToMovieDbContent(Content content, ContentType contentType) {
         if ( content == null ) {
             return null;
         }
 
         com.example.core.data.database.model.Content content1 = new com.example.core.data.database.model.Content();
 
         content1.setBackdropPath( content.getBackdropPath() );
         content1.setFirstAirDate( content.getFirstAirDate() );
         List<Integer list = content.getGenreIds();
         if ( list != null ) {
             content1.setGenreIds( new ArrayList<Integer( list ) );
         }
         content1.setId( content.getId() );
         content1.setName( content.getName() );
         List<String list1 = content.getOriginCountry();
         if ( list1 != null ) {
             content1.setOriginCountry( new ArrayList<String( list1 ) );
         }
         content1.setOriginalLanguage( content.getOriginalLanguage() );
         content1.setOriginalName( content.getOriginalName() );
         content1.setOverview( content.getOverview() );
         content1.setPopularity( content.getPopularity() );
         content1.setPosterPath( content.getPosterPath() );
         content1.setVoteAverage( content.getVoteAverage() );
         content1.setVoteCount( content.getVoteCount() );
 
         setContentType( content, content1, contentType );
 
         return content1;
     }
 
     @Override
     public void setContentType(Content content, com.example.core.data.database.model.Content dbContent, ContentType contentType) {
         if ( content == null ) {
             return;
         }
 
         dbContent.setBackdropPath( content.getBackdropPath() );
         dbContent.setFirstAirDate( content.getFirstAirDate() );
         if ( dbContent.getGenreIds() != null ) {
             List<Integer list = content.getGenreIds();
             if ( list != null ) {
                 dbContent.getGenreIds().clear();
                 dbContent.getGenreIds().addAll( list );
             }
             else {
                 dbContent.setGenreIds( null );
             }
         }
         else {
             List<Integer list = content.getGenreIds();
             if ( list != null ) {
                 dbContent.setGenreIds( new ArrayList<Integer( list ) );
             }
         }
         dbContent.setId( content.getId() );
         dbContent.setName( content.getName() );
         if ( dbContent.getOriginCountry() != null ) {
             List<String list1 = content.getOriginCountry();
             if ( list1 != null ) {
                 dbContent.getOriginCountry().clear();
                 dbContent.getOriginCountry().addAll( list1 );
             }
             else {
                 dbContent.setOriginCountry( null );
             }
         }
         else {
             List<String list1 = content.getOriginCountry();
             if ( list1 != null ) {
                 dbContent.setOriginCountry( new ArrayList<String( list1 ) );
             }
         }
         dbContent.setOriginalLanguage( content.getOriginalLanguage() );
         dbContent.setOriginalName( content.getOriginalName() );
         dbContent.setOverview( content.getOverview() );
         dbContent.setPopularity( content.getPopularity() );
         dbContent.setPosterPath( content.getPosterPath() );
         dbContent.setVoteAverage( content.getVoteAverage() );
         dbContent.setVoteCount( content.getVoteCount() );
 
         setContentType( content, dbContent, contentType );
     }
 }

```

I call the function "convertToMovieDbContent(content , contentType.MOVIE)"

I get a crash that logs : 
`019-02-20 17:44:41.906 10303-10303/com.example.guacamole W/System.err: io.reactivex.exceptions.UndeliverableException: java.lang.StackOverflowError: stack size 8MB
2019-02-20 17:44:41.913 10303-10303/com.example.guacamole W/System.err:     at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
2019-02-20 17:44:41.913 10303-10303/com.example.guacamole W/System.err:     at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:121)
2019-02-20 17:44:41.913 10303-10303/com.example.guacamole W/System.err:     at android.os.Handler.handleCallback(Handler.java:789)
2019-02-20 17:44:41.914 10303-10303/com.example.guacamole W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:98)
2019-02-20 17:44:41.914 10303-10303/com.example.guacamole W/System.err:     at android.os.Looper.loop(Looper.java:164)
2019-02-20 17:44:41.914 10303-10303/com.example.guacamole W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:6541)
2019-02-20 17:44:41.914 10303-10303/com.example.guacamole W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
2019-02-20 17:44:41.914 10303-10303/com.example.guacamole W/System.err:     at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
2019-02-20 17:44:41.915 10303-10303/com.example.guacamole W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
2019-02-20 17:44:42.023 10303-10303/com.example.guacamole W/System.err: Caused by: java.lang.StackOverflowError: stack size 8MB
2019-02-20 17:44:42.023 10303-10303/com.example.guacamole W/System.err:     at java.util.Arrays.copyOf(Arrays.java:3109)
2019-02-20 17:44:42.023 10303-10303/com.example.guacamole W/System.err:     at java.util.ArrayList.toArray(ArrayList.java:390)
2019-02-20 17:44:42.023 10303-10303/com.example.guacamole W/System.err:     at java.util.ArrayList.addAll(ArrayList.java:588)
2019-02-20 17:44:42.023 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:62)
2019-02-20 17:44:42.023 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:100)
2019-02-20 17:44:42.133 10303-10303/com.example.guacamole I/chatty: uid=10092(u0_a92) com.example.guacamole identical 1687 lines
2019-02-20 17:44:42.133 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:100)
2019-02-20 17:44:42.133 10303-10315/com.example.guacamole I/zygote: Background concurrent copying GC freed 112(53KB) AllocSpace objects, 0(0B) LOS objects, 19% free, 24MB/30MB, paused 96.736ms total 207.267ms
2019-02-20 17:44:42.133 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:100)
2019-02-20 17:44:43.352 10303-10303/com.example.guacamole I/chatty: uid=10092(u0_a92) com.example.guacamole identical 28495 lines
2019-02-20 17:44:43.352 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:100)
2019-02-20 17:44:43.352 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:100)
2019-02-20 17:44:43.352 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:100)
2019-02-20 17:44:45.856 10303-10303/com.example.guacamole I/chatty: uid=10092(u0_a92) com.example.guacamole identical 65535 lines
2019-02-20 17:44:47.132 10303-10303/com.example.guacamole I/chatty: uid=10092(u0_a92) com.example.guacamole identical 25459 lines
2019-02-20 17:44:47.133 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:100)
2019-02-20 17:44:47.133 10303-10315/com.example.guacamole I/zygote: Background concurrent copying GC freed 84062(6MB) AllocSpace objects, 0(0B) LOS objects, 21% free, 22MB/28MB, paused 1.283ms total 109.313ms
2019-02-20 17:44:47.133 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:100)
2019-02-20 17:44:47.543 10303-10303/com.example.guacamole I/chatty: uid=10092(u0_a92) com.example.guacamole identical 7847 lines
2019-02-20 17:44:47.543 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.setContentType(ContentMapperImpl.java:100)
2019-02-20 17:44:47.543 10303-10303/com.example.guacamole W/System.err:     at com.example.core.data.mapper.ContentMapperImpl.convertToMovieDbContent(ContentMapperImpl.java:45
`

**Comment by sjaakd** [comment]
So.. did you try to do some debugging? E.g. put a breakpoint in the `setContentType`? How does the call stack look like?

Since it does not look like a MapStruct issue to me, I'll provide an answer on SO..

**Comment by sjaakd** [comment]
Like said, its not a MapStruct issue. @filiphr : do you see a need to support these Kotlin cases?

**Comment by filiphr** [comment]
There is nothing that we can do from the Java side. This is a known problem in Kotlin. Have a look at https://github.com/mapstruct/mapstruct/issues/1577. Kotlin as `@JvmDefault` annotation but that also doesn't work (see [KT-25960](https://youtrack.jetbrains.com/issue/KT-25960)). For the time being I would suggest that you use an abstract class instead of an interface.

**Comment by martinweu** [comment]
According to the updated issue [KT-25960](https://youtrack.jetbrains.com/issue/KT-25960) a fix will be provided in 1.4.20.

**Comment by filiphr** [comment]
Thanks for the info @martinweu. To add to that, there will be no need for any change on the MapStruct side, which means that once you upgrade your Kotlin to 1.4.20 MapStruct will see the modifiers and won't provide implementations for those methods.
